multi-line error messages

kedashoe opened the pull request:
Before
```
TypeError: ‘aa’ expected a value of type Number as its second argument; received /x/
```
After
```
TypeError: ‘aa’ expected its first and second arguments to be of the same type; the first is of type Number, however the second, /x/, is of type RegExp
```
I realize one might normally file an issue first, but was just playing around with the code and figured may as well submit a PR, no hard feelings if you don't want to go down this path. Couple notes

- When the placeholder is used, the message is a little funny (see https://github.com/kedashoe/sanctuary-def/blob/same-type-var-error/test/index.js#L974-L976). I was going to address it but left it as it was as it actually conveys a little more information
- This hit 8 tests, so I didn't add any new ones
- The pattern `commonTypes(map(values, determineActualTypes))` is now used 4 times, and I believe this is the only time `determineActualTypes` is used, but I just reused the pattern rather than refactor
- I liked `however` in the error message at first, but don't anymore. I feel like maybe there is a really nice multi-line error message to be arrived at somewhere here. (And not to hijack my own PR, but what do you all think about using `1st`, `2nd`... rather than `first`, `second`...? to me that is much easier to read and quickly process)

Great work on the lib :)

kedashoe commented:
Couple more notes

- the error messages could be improved for unary and binary types, I wasn't sure exactly how to word it though
- I'd need to update the docs in some places I think
- looks like ci is mad but I can't tell what is going wrong? `lint` and `test` appear to be happy on my machine

davidchambers commented:
Interesting pull request, @kedashoe! I'll review it in the morning, but I have one initial thought.

> ```javascript
> '‘concat’ expected its first and second arguments to be ' +
> 'of the same type; the first is of type String, however ' +
> 'the second, Left([1, 2, 3]), is of type ' +
> '(Either (Array Number) ???)'
> ```

This format will be awkward when a value belongs to multiple types. We'd need to enumerate the types.

We could write _share a type_ to avoid implying that each value belongs to just one type.

How about this?

_‘concat’ expected its first and second arguments to share a type; Left('abc') and Left([1, 2, 3]) do not_

davidchambers commented:
This is the reason for the test failure:

>     ERROR: Coverage for branches (98.7%) does not meet global threshold (100%)

100% branch coverage is necessary—but not sufficient—to ensure the library is thoroughly tested. :)

After running `make test` one can run `open coverage/lcov-report/index.html` to view the coverage report.

davidchambers commented on test/index.js:
```diff
@@ -616,8 +616,9 @@ describe('def', function() {
-                   'as its second argument; received "XXX"'));
+                   '‘defaultTo’ expected its first and second arguments to ' +
+                   'be of the same type; the first is of type Number' +
+                   ', however the second, "XXX", is of type String'));
```
This is not quite correct, since `defaultTo :: a -> Nullable a -> a`.

kedashoe replied:
If I understand correctly, this is what I meant by

> the error messages could be improved for unary and binary types, I wasn't sure exactly how to word it though

Do you have any thoughts on how to say this nicely? Maybe something like

`‘defaultTo’ expected its first and second arguments to ??? (derive from the same type? be of the same base type?); the first is of type Number, so the second was expected to be of type (Nullable Number) but received "XXX", which is of type String`

davidchambers replied:
Let's consider a function such as [`S.fromMaybe :: a -> Maybe a -> a`][1], because I find reasoning about `Nullable a` difficult due to Nullable not being a container.

It's incorrect to state that the first and second arguments must share a type. There is a constraint, but it's not easy to express in words that the value inside the Maybe must share a type with the value provided as the first argument. Perhaps the most succinct approach is to present that type signature and state that the `a`s don't match. We could elaborate by listing the types of each `a` value and stating that the intersection between these sets is empty.

For example:

```
> S.fromMaybe(0, S.Just('XXX'))
TypeError: TypeVariable constraint violation

fromMaybe :: a -> Maybe a -> a
             ^          ^
             1          2

The value at position 1 is 0, which is a member of the following types:

  - Number
  - sanctuary-def/ValidNumber
  - sanctuary-def/Integer

The value at position 2 is "XXX", which is a member of the following
types:

  - String

Since there is no type of which both values are members, the
type-variable constraint has been violated.
```

Now *this* is an error message that would fill me with hope for our profession. :)


[1]: https://github.com/plaid/sanctuary#fromMaybe

kedashoe replied:
> but it's not easy to express in words

Agreed! I like what you've done there, I'll play around a bit and see what I can come up with.

kedashoe replied:
Alright, pushed some changes. If you pull down the branch and run `make test`, `supports "nullable" types` should fail and you should see our new error message. Most of the diff is creating a `signature` POJO that we can in theory use to print our signature and highlight certain parts (eg `sigShowTypeVarPositions`). I pretty much just tried to make as few changes as possible in this iteration to surrounding code. The implementation is mostly imperative.

The signature is made up of four parts (name, constraints, arguments, return type) which I generically have called "group"s. Each group can have 0 or more "item"s (ie a function that takes 2 arguments would have 2 items)

I haven't implemented constraints. I didn't really look into it, but I wasn't sure off the top of my head how we could derive something like `filter :: (Monad m, Monoid (m a)) => (a -> Boolean) -> m a -> m a`

Also, `_satisfactoryTypes` now takes 10 arguments, so we have that going for us :dancer: I think it might be nice to create a POJO in `def` that simply wraps the things the signature needs and pass that around, ie

```js
return curry(name,
             constraints,
             expArgTypes,
             expTypes[expTypes.length - 1],
             {},
             new Array(arity),
             range(0, arity),
             impl);
```
would become
```js
var _sig = {
  name: name,
  constraints: constraints,
  args: expArgTypes,
  ret: expTypes[expTypes.length - 1]
};
return curry(_sig, {}, new Array(arity), range(0, arity), impl);
```
I'm sure I had more to say about this all, but it's getting late :sleeping: Let me know what you think of this direction.

davidchambers replied:
I pulled down your branch and tried it. Fantastic stuff!!!

davidchambers commented on test/index.js:
```diff
@@ -627,8 +628,9 @@ describe('def', function() {
-                   '(Nullable String); returned 42'));
+                   '‘f’ expected its first argument and return value to be ' +
+                   'of the same type; the first argument is of type String' +
+                   ', however the return value, 42, is of type Number'));
```
We appear to be comparing the `a`s only. The first argument is really of type `Nullable String` in this context.

kedashoe commented:
> 100% branch coverage is necessary—but not sufficient—to ensure the library is thoroughly tested. :)
> 
> After running make test one can run open coverage/lcov-report/index.html to view the coverage report.

Ah, thank you, I'll look into this

davidchambers commented on index.js:
```diff
@@ -608,6 +794,24 @@
+      );
+    }
+  };
+
```
Here's a slightly improved version:

```javascript
//  ordinalList :: [Integer] -> String
var ordinalList = function(indexes) {
  var words = map(indexes, function(index) { return ordinals[index]; });
  return words.length > 2 ?
    words.slice(0, -1).join(', ') + ', and ' + words[words.length - 1] :
    words.join(' and ');
};
```

This gives us my beloved Oxford comma. :)

kedashoe replied:
I'm not sure we'll end up needing this function with the new error messages, but I've updated it in the meantime

davidchambers commented on index.js:
```diff
@@ -125,6 +145,172 @@
+  var sigGroupStr = function(sigPart) {
+    return (
+      isEmpty(sigPart.items) ? '' :
+      map(sigPart.items, x => x.text).join(sigPart.glue) + sigPart.separator
```
Let's stick to ES3 language features. ;)

kedashoe replied:
ha! didn't even realize I did that

davidchambers commented on index.js:
```diff
@@ -125,6 +145,172 @@
+      r += s;
+    }
+    return r;
+  };
```
I prefer this implementation:

```javascript
var strRepeat = function(s, times) {
  return Array(times + 1).join(s);
};
```

davidchambers commented on index.js:
```diff
@@ -651,6 +855,45 @@
+      'member of the following types:\n\n' +
+      ul(types2) + '\n\n' +
+      'Since there is no type of which both values are members, the ' +
+      'type-variable constraint has been violated.\n'
```
Let's drop the trailing `'\n'`.

kedashoe replied:
I originally did not have the trailing newline, but I felt like it looked better with one in there. Did you try it both ways (also outside of mocha, which colors the output)?

davidchambers replied:
I hadn't tried it. I just did so, and I agree that an empty line between the last line of our error message and the rest of the stack trace is very nice indeed. Do we have any other multi-line messages? If so, let's append `'\n'` to each.

davidchambers commented on index.js:
```diff
@@ -125,6 +145,172 @@
+      }
+    }
+    return item;
+  };
```
Could you explain this function in a few sentences? I'd like to think we can simplify it quite a bit. :)

kedashoe replied:
Recursively creating the argument [items](https://github.com/plaid/sanctuary-def/pull/23/files#diff-168726dbe96b3ce427e7fedce31bb0bcR160). I was planning to change this to make two passes, one to create the text and another to compute the offsets.. thoughts? We hardly need to worry about perf here at all since the sig structure is only built once and only when there is an error to be thrown. We could even make multiple passes just to create the text, ie don't work about wrapping children in parentheses as we go and do it in a 2nd pass.

davidchambers commented on index.js:
```diff
@@ -651,6 +855,45 @@
+    return isNaN(index) ?
+      conflictingTypeVarReturnValue(sig, type, value1, prevTypes) :
+      conflictingTypeVarArgument(sig, name, index, value1, types1, value2, types2);
+  };
```
I think we could use the same approach for both arguments and return values.

```
fromMaybe :: a -> Maybe a -> a
             ^          ^
             1          2
```

```
fromMaybe :: a -> Maybe a -> a
             ^               ^
             1               2
```

I'm guessing you haven't yet had a chance to work on the handling of invalid return values.

kedashoe replied:
Ya, haven't updated the return value error message yet

davidchambers commented on index.js:
```diff
@@ -101,6 +121,13 @@
     };
   };
+  
+  //  prop :: String -> Obj -> Any
```
`s/Obj/Object/`

davidchambers commented on index.js:
```diff
@@ -92,6 +97,21 @@
+      }
+      for (k in y) {
+        z[k] = y[k];
+      }
```
I *think* the linter will tolerate this change:

```diff
-      for (k in x) {
-        z[k] = x[k];
-      }
-      for (k in y) {
-        z[k] = y[k];
-      }
+      for (k in x) z[k] = x[k];
+      for (k in y) z[k] = y[k];
```

davidchambers commented on index.js:
```diff
@@ -125,6 +152,168 @@
+      );
+    };
+    return sig;
+  };
```
This gets us some of the way there, in simpler fashion:

```javascript
var signature = function(name, constraints, expTypes, expRetType) {
  var constraintsRepr = '';
  var varNames = keys(constraints).sort();
  if (varNames.length > 0) {
    var reprs = map(varNames, function(varName) {
      return map(constraints[varName], function(typeClass) {
        return String(typeClass) + ' ' + varName;
      });
    });
    constraintsRepr = '(' + reprs.join(', ') + ') => ';
  }
  return name + ' :: ' + constraintsRepr +
         expTypes.concat([expRetType]).join(' -> '));
};
```

Limitations:

  - Does not preserve the tokens, making it difficult to determine the locations of type variables. I'll work on a token-based approach.

  - Parens are used unnecessarily in some cases. I'm fine with this. It simplifies things significantly, and it's consistent with the current string representations of "container" types. If we see a straightforward way to avoid unnecessary parens, though, I'm in favour of doing so. :)

kedashoe replied:
I'll explain a little the signature "item" I arrived at:
```js
var sigItem = function(type, offset, text, glue) {
  return {
    type: type,
    text: text,
    glue: glue,
    start: offset,
    end: offset + text.length,
    children: [],
    typeVars: {}
  };
};
```
**type:** right now this is to just to help with [building the data structure](https://github.com/plaid/sanctuary-def/pull/23/files#diff-168726dbe96b3ce427e7fedce31bb0bcR239), although it could also be used in error messages
**text:** the text
**glue:** help with displaying
**start and end:** so far we are only highlighting type variables. I thought it might be nice to convert all error messages (for which it makes any sense) to this multi-line format, and for some of those we might want to highlight the entire container type, eg
```js
fromMaybe :: a -> Maybe a -> a
                  ^^^^^^^
                     1
```
I was going to do a similar thing with constraints
**children:** I nested container types mostly just to mimic how they are actually defined. Also each item has its own dict of typeVars. I wasn't sure if a definition such as `BinaryX a (BinaryY a b)` was possible, in which `a` is in two places in the same argument. I have assumed `BinaryX a a` is never used, but the dict is just for lookup convenience/speed, we can change to an array and `find` things if we need to
**typeVars:** easy way to find typeVars and their offsets

davidchambers replied:
I've been working on an approach which seems simpler to me.

```javascript
//  literal :: String -> Token
var literal = function(text) {
  return {type: 'LITERAL', text: text};
};

//  variable :: String -> Token
var variable = function(name) {
  return {type: 'VARIABLE', name: name};
};

//  group :: String,String,String -> [Token] -> Token
var group = function(open, separator, close) {
  return function(tokens) {
    return {type: 'GROUP',
            open: open,
            separator: separator,
            close: close,
            tokens: tokens};
  };
};

var EMPTY = '';
var SPACE = ' ';

var constraints = group('(', ', ', ')');
var constraint  = group(EMPTY, SPACE, EMPTY);
var types       = group(EMPTY, ' -> ', EMPTY);
var type        = group('(', SPACE, ')');

//  serialize :: Token -> String
var serialize = function recur(token) {
  switch (token.type) {
    case 'LITERAL':
      return token.text;
    case 'VARIABLE':
      return token.name;
    default:
      return token.open +
             token.tokens.map(recur).join(token.separator) +
             token.close;
  }
};

//  token :: Token
var token =
group(EMPTY, ' :: ', EMPTY)([
  literal('Either.concat'),
  group(EMPTY, ' => ', EMPTY)([
    constraints([constraint([literal('Semigroup'), variable('a')]),
                 constraint([literal('Semigroup'), variable('b')])]),
    types([type([literal('Either'), variable('a'), variable('b')]),
           type([literal('Either'), variable('a'), variable('b')]),
           type([literal('Either'), variable('a'), variable('b')])]),
  ]),
]);

token;
// => { type: 'GROUP',
//      open: '',
//      separator: ' :: ',
//      close: '',
//      tokens:
//       [ { type: 'LITERAL', text: 'Either.concat' },
//         { type: 'GROUP',
//           open: '',
//           separator: ' => ',
//           close: '',
//           tokens:
//            [ { type: 'GROUP',
//                open: '(',
//                separator: ', ',
//                close: ')',
//                tokens:
//                 [ { type: 'GROUP',
//                     open: '',
//                     separator: ' ',
//                     close: '',
//                     tokens:
//                      [ { type: 'LITERAL', text: 'Semigroup' },
//                        { type: 'VARIABLE', name: 'a' } ] },
//                   { type: 'GROUP',
//                     open: '',
//                     separator: ' ',
//                     close: '',
//                     tokens:
//                      [ { type: 'LITERAL', text: 'Semigroup' },
//                        { type: 'VARIABLE', name: 'b' } ] } ] },
//              { type: 'GROUP',
//                open: '',
//                separator: ' -> ',
//                close: '',
//                tokens:
//                 [ { type: 'GROUP',
//                     open: '(',
//                     separator: ' ',
//                     close: ')',
//                     tokens:
//                      [ { type: 'LITERAL', text: 'Either' },
//                        { type: 'VARIABLE', name: 'a' },
//                        { type: 'VARIABLE', name: 'b' } ] },
//                   { type: 'GROUP',
//                     open: '(',
//                     separator: ' ',
//                     close: ')',
//                     tokens:
//                      [ { type: 'LITERAL', text: 'Either' },
//                        { type: 'VARIABLE', name: 'a' },
//                        { type: 'VARIABLE', name: 'b' } ] },
//                   { type: 'GROUP',
//                     open: '(',
//                     separator: ' ',
//                     close: ')',
//                     tokens:
//                      [ { type: 'LITERAL', text: 'Either' },
//                        { type: 'VARIABLE', name: 'a' },
//                        { type: 'VARIABLE', name: 'b' } ] } ] } ] } ] }

serialize(token);
// => 'Either.concat :: (Semigroup a, Semigroup b) => (Either a b) -> (Either a b) -> (Either a b)'
```

Outstanding tasks:

  - Avoid including `'() => '` when there are no constraints. I believe a general solution exists.
  - Generate the `^` markers from the same data structure.
  - Write a function to generate a value of type `Token` from `(name, constraints, expTypes, expRetType)`.

davidchambers replied:
I've made some improvements to the code I posted above. Incorporate into your patch as much or as little of this code as your deem appropriate, @kedashoe. :)

kedashoe replied:
Alright sounds good :) The weekend is approaching, I'll check in/work on this when I can but time will probably be limited :baby:

davidchambers replied:
No pressure at all. :baby: > :ram:, but hopefully :ram: > :tv:. :)

kedashoe replied:
haha.. no TV at the moment so it's just :ram: vs :book:

kedashoe commented:
New version up, here is what the object representing a signature looks like now, using `fromMaybe`. 

```js
{ type: 'SIG',
  text: '',
  start: 0,
  end: 30,
  tokens: 
   [ { type: 'NAME',
       text: '',
       start: 0,
       end: 13,
       tokens: 
        [ { type: 'FN_NAME',
            text: 'fromMaybe',
            start: 0,
            end: 9,
            tokens: [] },
          { type: 'DELIMITER', text: ' :: ', start: 9, end: 13, tokens: [] } ] },
     { type: 'CONSTRAINTS', text: '', start: 13, end: 13, tokens: [] },
     { type: 'ARGS',
       text: '',
       start: 13,
       end: 29,
       tokens: 
        [ { type: 'VARIABLE', text: 'a', start: 13, end: 14, tokens: [] },
          { type: 'DELIMITER', text: ' -> ', start: 14, end: 18, tokens: [] },
          { type: 'UNARY',
            text: 'Maybe',
            start: 18,
            end: 25,
            tokens: 
             [ { type: 'DELIMITER', text: ' ', start: 23, end: 24, tokens: [] },
               { type: 'VARIABLE', text: 'a', start: 24, end: 25, tokens: [] } ] },
          { type: 'DELIMITER', text: ' -> ', start: 25, end: 29, tokens: [] } ] },
     { type: 'RET',
       text: '',
       start: 29,
       end: 30,
       tokens: [ { type: 'VARIABLE', text: 'a', start: 29, end: 30, tokens: [] } ] } ]
```
Same object all the way down. I've also tested it with
```js
def('layers', {}, [a, Either(b, Either(a, Either(Either(c, d), b))), $.Any], function(x, either) {
  return 1;
});
```
and everything seemed good. Will make additional comments inline.

kedashoe commented on index.js:
```diff
@@ -124,6 +151,298 @@
+  // closure.. no reason really beyond helping with name clashes
+  var signature = (function(){
+    
+    var Grammar = function(open, close, prefix, glue, separator) {
```
As I find is often the case with configuration, this seemed like a good idea at first but I'm not sure there is much use it. Pretty much everything is a special case, and the functions would be so simple anyway I don't know that this is the most useful way to define things.

kedashoe commented on index.js:
```diff
@@ -124,6 +151,298 @@
+    }
+    
+    //  Token :: (String, String, [Token]) -> Token
+    var Token = function(type, text, children) {
```
Everything in the signature is a `token`

kedashoe commented on index.js:
```diff
@@ -124,6 +151,298 @@
+        Constraints(constraints),
+        Args(expTypes),
+        Ret(expRetType),
+      ]);
```
First we create tokens based on user input. At this point if we printed out our sig it would look like `fromMaybeaMaybeaa`.

kedashoe commented on index.js:
```diff
@@ -124,6 +151,298 @@
+      traverse(addPrefix, null, sig);
+      traverse(addWrapper, null, sig);
+      traverse(addSeparator, null, sig);
+      pairTraverse(addGlue, null, sig);
```
These all mutate our tree, adding in the delimiters.

kedashoe commented on index.js:
```diff
@@ -124,6 +151,298 @@
+      
+      // metrics
+      traverse(setStart, 0, sig);
+      postTraverse(setEnd, 0, sig);
```
Set `start` and `end` position of all our tokens

kedashoe commented on index.js:
```diff
@@ -124,6 +151,298 @@
+  //  traverse :: (((a, b, b) -> a) -> a -> [b] -> a
+  var traverse = function(f, acc, node, parent) {
+    acc = f(acc, node, parent);
+    var tokens = arrayClone(node.tokens);
```
Much easier to reason about our mutating traversals if we don't have to worry about the stuff we are adding as we go.

davidchambers commented on index.js:
```diff
@@ -373,86 +692,62 @@
 
   //  ValidDate :: Type
-  $.ValidDate = NullaryType(
+  $.ValidDate = $.NullaryType(
```
Let's revert these prefix changes.

davidchambers commented on index.js:
```diff
@@ -124,6 +151,298 @@
+    var insert = function(children, target, newChild, offset) {
+      for (var idx = 0; idx < children.length; idx += 1) {
+        if (children[idx] === target) {
+          children.splice(idx + offset, 0, newChild);
```
Could we avoid mutation here?

kedashoe replied:
I'm sure that it is possible ;) Don't know how easy/difficult though. I made a comment elsewhere related to this.. we create a cloned array as we traverse, which simplifies things I think. I'm not really sure what we would gain, we would just end up changing
```js

      traverse(addPrefix, null, sig);
      traverse(addWrapper, null, sig);
      traverse(addSeparator, null, sig);
      pairTraverse(addGlue, null, sig);
```
to
```js
      sig = traverse(addPrefix, null, sig);
      sig = traverse(addWrapper, null, sig);
      sig = traverse(addSeparator, null, sig);
      sig = pairTraverse(addGlue, null, sig);
```
or a nicer looking version that uses a pipe of some sort to achieve the same thing. If we get everything to a place we like and this becomes a sticking point, I can revisit it. I know babel plugins mutate their AST as they traverse. And sanctuary-def does some mutation already as well https://github.com/plaid/sanctuary-def/blob/master/index.js#L762

kedashoe commented on index.js:
```diff
@@ -124,6 +151,298 @@
+  };
+  
+  //  postTraverse :: (((a, b, b) -> a) -> a -> [b] -> a
+  var postTraverse = function(f, acc, node, parent) {
```
Still depth first, but we call our accumulating function after recurring so that we start at the leaves.

kedashoe commented on index.js:
```diff
@@ -124,6 +151,298 @@
+  };
+
+  //  pairTraverse :: (((a, b, b, b) -> a) -> a -> [b] -> a
+  var pairTraverse = function(f, acc, node) {
```
This is currently used by `addGlue`. Having this amounts to two 'simple' functions vs using `traverse` for `addGlue` and making that function more complicated.

kedashoe commented:
New version up, if you run tests you should see a couple errors, including the constraint error message. Sorry for short message, gotta run, will check back later :runner:

davidchambers commented:
> New version up, if you run tests you should see a couple errors, including the constraint error message.

I pulled and ran the tests. Nice progress!

kedashoe commented:
New error message for invalid values and added a temporary `describe.only` block to the test file for all this work. Currently we have

1. conflicting type var args
2. conflicting type var arg/return value
3. unmet arg constraint
4. unmet return value constraint (I don't believe this was tested previously, and it could potentially come up as constraint violation is tested before type var conflicts)
5. type checks its args
6. sig wraps nested containers in parens (just to test that signatures are properly being represented. would like to add more of these)

kedashoe commented:
Now type checking return value with new error message. I think that is a decent group for a first commit. We could also include the `orphanArgument` error.

I was thinking about the error messages, what do you think about something like this:

Invalid value:
```
TypeError: Invalid value

div :: ValidNumber -> NonZeroInteger -> ValidNumber
                      ^^^^^^^^^^^^^^
                            1

At position 1:
  Where: 2nd Argument
  Value: 0
  Types: Number, ValidNumber

Expected types: NonZeroInteger.
```
Type variable conflict:
```
TypeError: Type variable conflict

aa :: a -> a -> Pair a a
      ^    ^
      1    2

At position 1:
  Where: 1st Argument
  Value: 0
  Type: Number

At position 2:
  Where: 2nd Argument
  Value: /x/
  Type: RegExp

Since there is no type of which both values are members, the type-variable constraint has been violated.
```
Constraint violation:
```
TypeError: Constraint violation

or :: Alternative a => a -> a -> a
      ^^^^^^^^^^^^^              ^
                                 1

At position 1:
  Where: Return Value
  Value: Left(1)
  Types: (Either Number ???), (Either Integer ???)
  
‘or’ requires ‘a’ to implement Alternative.
```
I think this is nice as at a very quick glance you get a lot of info (added benefit, we don't have to do as much writing, and the wording can get difficult (I arrived at this after trying to think of a way to include which argument was the offender in our error messages)). Let me know what you think.

All the new functions should have type signatures now as well.

davidchambers commented:
The proposed format looks good to me, on the whole!

I'm not sure the `Where:` field is a good idea, though. It works in this case:

> ```
> TypeError: Type variable conflict
>
> aa :: a -> a -> Pair a a
>       ^    ^
>       1    2
>
> At position 1:
>   Where: 1st Argument
> [snip]
> ```

How would we handle a function of type `[a] -> [a] -> [a]`, though? `Where: Somewhere inside the first argument`? :stuck_out_tongue_winking_eye:

The position markers already do a good job of pinpointing offenders, so we could simply drop `Where:`.

> ```
> At position 1:
>   Where: 1st Argument
>   Value: 0
>   Type: Number
>
> At position 2:
>   Where: 2nd Argument
>   Value: /x/
>   Type: RegExp
> ```

If we're to drop `Where:`, we should consider a more compact format such as:

```
1)  0 :: Number

2)  /x/ :: RegExp
```

davidchambers commented on index.js:
```diff
@@ -833,9 +1408,18 @@
           if (!expRetType.test(returnValue)) {
-            throw invalidReturnValue(name, [expRetType], returnValue);
+            throw invalidValue(signature(name,
+                                               constraints,
```
:triangular_ruler:

kedashoe commented:
> Where: Somewhere inside the first argument? :stuck_out_tongue_winking_eye:

haha.. this is actually exactly what I had in mind! I was thinking about adding the argument to the description.. but if you have
```
x :: Either a b -> a
            ^
            1
```
saying "The first argument, at position 1, has a value of 0, which is a member of the following types..." seemed a bit off, so then you have to check if it a container type and use different wording or what have you. I feel like `Where: 1st argument` is correct enough and provides useful information: I know which argument in my function is causing the problem. Obviously you can figure it out from the carrot/position number, but for more complicated signatures it is nice and I don't think it is hurting anything.

davidchambers commented:
I may be missing your point.

I was thinking that for `concat :: [a] -> [a] -> [a]` we could differentiate the following errors:

```javascript
concat([1, 2, 3], 'XXX');
// ! TypeError: Invalid value
//
// concat :: Array a -> Array a -> Array a
//                      ^^^^^^^

concat([1, 2, 3], ['XXX']);
// ! TypeError: Constraint violation
//
// concat :: Array a -> Array a -> Array a
//                            ^
```

In the first case the second argument in (completely) invalid. In the second case the second argument is (partially) invalid, but we can be more specific when indicating the location of the error.

kedashoe commented:
> In the first case the second argument in (completely) invalid. In the second case the second argument is (partially) invalid, but we can be more specific when indicating the location of the error.

I may be poorly conveying my point ;) I think you've got it though, I'm just saying that in the second case
```
// concat :: Array a -> Array a -> Array a
//                            ^
```
even though "2nd argument" does not describe *exactly* where the problem is, it is additional information that will be useful to developers, and it is not incorrect. In both cases, you would have "Where: 2nd Argument", but the carrots would be different (as you indicated.. and the algorithm does currently produce the carrots as you've indicated). If you ask me where I am, I could answer simply "Aliso Village", but I could also answer "Aliso Village, Los Angeles"; I am arguing for the latter.

davidchambers commented:
> ```javascript
> concat([1, 2, 3], ['XXX']);
> // ! TypeError: Constraint violation
> //
> // concat :: Array a -> Array a -> Array a
> //                            ^

In this case, would we display `Value: ["XXX"]` or `Value: "XXX"`?

kedashoe commented:
Updated error messages, fixed most of the linting errors, added files for error messages to aid discussion.

> In this case, would we display Value: ["XXX"] or Value: "XXX"?

Hmm, I'm not sure. And pertinent whether or not we have the `Where: ` clause right? We still have this question with our highlighting of that `a`. When we have an array, it's actually just a sample value. See `test/error-messages/nested-type-var-conflict.txt` for this one as it currently stands.

davidchambers commented:
Thanks for adding the error message examples as separate files. This will make it easier to discuss the various messages concurrently without getting confused. :)

davidchambers commented on test/error-messages/constraint-arg.txt:
```diff
@@ -0,0 +1,12 @@
+  Value: /x/
+  Types: RegExp
+
+Expected types: (Array a).
```
The expected type, in this case at least, reiterates the underlined portion of the type signature. Will this always be the case? If not, could you provide an example?

I favour something more minimal:

    Invalid value

    concat :: Array a -> Array a -> Array a
              ^^^^^^^
                 1

    1)  /x/ :: RegExp

    The value at position 1 is not a member of (Array a).

davidchambers commented on test/error-messages/invalid-arg-value.txt:
```diff
@@ -0,0 +1,12 @@
+  Value: null
+  Types: Null
+
+Expected types: Number.
```
Invalid value

    sqrt :: Number -> Number
            ^^^^^^
              1

    1)  null :: Null

    The value at position 1 is not a member of Number.

kedashoe replied:
All error messages updated to use this format. I removed the `.only` from the describe block I added, so you can see a bunch more error messages if you run the test suite.

davidchambers commented on index.js:
```diff
@@ -125,6 +140,473 @@
+    return xs[xs.length - 1];
+  };
+
+  //  toPairs :: {k: v} -> [[k, v]]
```
I'd prefer this type signature:

```haskell
toPairs :: StrMap a -> [Pair String a]
```

davidchambers commented on index.js:
```diff
@@ -125,6 +140,473 @@
+    return b;
+  };
+
+  // closure.. no reason really beyond helping with name clashes
```
What are the name clashes specifically? Let's remove this IIFE if we can. :)

davidchambers commented on index.js:
```diff
@@ -646,28 +1128,82 @@
+  //   '7th',
+  //   '8th',
+  //   '9th'
+  // ];
```
Remove this?

davidchambers commented on test/error-messages/constraint-arg.txt:
```diff
@@ -0,0 +1,9 @@
+
+1) 42 :: Number
+
+‘semiFn’ requires ‘a’ to implement Semigroup.
```
What do you think of this tweak?

```diff
-‘semiFn’ requires ‘a’ to implement Semigroup.
+‘semiFn’ requires ‘a’ to satisfy the Semigroup type-class constraint.
+The value at position 1 does not satisfy the constraint, therefore
+‘a’ does not satisfy the constraint.
```

On a related note, do you think we should hard-wrap long lines? It would make messages easier to read in a wide terminal. Were we to wrap at 79 characters, say, we'd risk unfortunate line wrapping in a slightly narrower terminal.

I'm slightly in favour of hard wrapping.

kedashoe replied:
> What do you think of this tweak?
> 
> -‘semiFn’ requires ‘a’ to implement Semigroup.
> +‘semiFn’ requires ‘a’ to satisfy the Semigroup type-class constraint.
> +The value at position 1 does not satisfy the constraint, therefore
> +‘a’ does not satisfy the constraint.

Sure, nice for newer users to be explicit and the explanation won't hurt the veterans.

> On a related note, do you think we should hard-wrap long lines? It would make messages easier to read in a wide terminal. Were we to wrap at 79 characters, say, we'd risk unfortunate line wrapping in a slightly narrower terminal.
> 
> I'm slightly in favour of hard wrapping.

Personally I never hard wrap anything, so I guess I am in favor of leaving it :wink: Not unreasonable as its own PR, so maybe you can see how it looks as it is and then add a `wrap` function later?

davidchambers replied:
> Not unreasonable as its own PR, so maybe you can see how it looks as it is and then add a `wrap` function later?

Works for me. :)

davidchambers commented on test/error-messages/constraint-arg.txt:
```diff
@@ -0,0 +1,9 @@
+Constraint violation
```
Let's be more specific:

```diff
-Constraint violation
+Type-class constraint violation
```

davidchambers commented on test/error-messages/lots-of-eithers.txt:
```diff
@@ -0,0 +1,10 @@
+Type-variable conflict
```
I'd prefer "Type-variable constraint violation".

davidchambers commented on test/error-messages/nested-invalid-value.txt:
```diff
@@ -0,0 +1,9 @@
+
+1) "ab" :: String
+
+The value at position 1 is not a member of (Array Number).
```
This message does not explain how we determined `a` to be `Number`. Ideally I'd like to see something like this:

    Invalid value

    concat :: Array a -> Array a -> Array a
              ^^^^^^^    ^^^^^^^
                 1          2

    1)  [1, 2] :: Array Number
    2)  "ab" :: String

    The value at position 2 is not a member of (Array Number).

I realize this may be tricky. It's very much a nice-to-have rather than a must-have.

kedashoe replied:
This is in contrast to `nested-type-var-conflict`
```
Type-variable conflict

concat :: Array a -> Array a -> Array a
                ^          ^
                1          2

1) 1 :: Number

2) "a" :: String

Since there is no type of which both values are members, the type-variable constraint has been violated.
```
Since it doesn't matter what type `a` is `nested-invalid-value` as "ab" will always be in violation, I wonder if it would be better to use `???` as in:
```
Invalid value

concat :: Array a -> Array a -> Array a
                     ^^^^^^^
                        1

1) "ab" :: String

The value at position 1 is not a member of (Array ???).
```
Currently we don't pass multiple positions to `invalidValue`, so it would be a bit of work to show both. I haven't considered `???` instead of the actual value, not sure off the top of my head how difficult it would be to determine when that would be appropriate.

davidchambers replied:
Using `???` here is a good idea. Let's go with that. :)

kedashoe replied:
:zap:ed all the rest, will look into this one.

Also I went back to including a blank line between the position detail lines.. it feels like an extravagant amount of white space so I want to like it better without, but I don't :sheep:

davidchambers replied:
The changes look great to me!

davidchambers commented on index.js:
```diff
@@ -646,28 +1118,64 @@
+
+  //  showVal :: (Integer, a, [Type]) -> String
+  var showVal = function(num, value, types) {
+    return num + ') ' + show(value) + ' :: ' + types.join(', ') + '\n\n';
```
I'd like to have two spaces after the `)`. I'd like the value and the type to be more tightly bound visually than the value and the number.

One space:

    1) 1 :: Number

    2) "a" :: String

Two spaces:

    1)  1 :: Number

    2)  "a" :: String

davidchambers commented on index.js:
```diff
@@ -125,6 +140,463 @@
+
+  //  showSignature :: Node -> String
+  var showSignature = function(tree) {
+    return traverseWith(function(acc, t) { return acc.concat(t.text); },
```
What's the type of `acc` here? If `String`, let's use `acc + t.text`. If `Array a`, let's use `acc.concat([t.text])`.

kedashoe replied:
yep, :zap:, and zapped two spaces after `)`

davidchambers commented on index.js:
```diff
@@ -125,6 +136,463 @@
+      children: children
+    };
+
+    return t;
```
Let's return the value directly.

davidchambers commented on index.js:
```diff
@@ -125,6 +136,463 @@
+
+  //  arrayClone :: [a] -> [a]
+  var arrayClone = function(a) {
+    return a.slice(0);
```
Let's omit the start index here as well.

davidchambers commented on index.js:
```diff
@@ -125,6 +136,463 @@
+  //  getParent :: [Node] -> Nullable Node
+  var getParent = function(parents) {
+    return last(parents);
+  };
```
Does this function add value? Could we simply use `last` directly?

davidchambers commented on index.js:
```diff
@@ -125,6 +136,463 @@
+    return last(parents);
+  };
+
+  //  head :: [a] -> Nullable a
```
Strictly speaking the return type is not `Nullable a`, since the Nullable type differentiates between `null` and `undefined`. We could change the function's body to `return xs.length > 0 ? xs[0] : null;`, or change the type signature to something vague such as `head :: [a] -> a?`.

davidchambers commented on index.js:
```diff
@@ -125,6 +136,463 @@
+  //  isRoot :: [Node] -> Boolean
+  var isRoot = function(parents) {
+    return isEmpty(parents);
+  };
```
We could define `isRoot` more simply: `var isRoot = isEmpty;`. Do you think this alias is worthwhile?

kedashoe replied:
I think `isRoot` (`getParent`, etc) is nice. They especially help with refactoring as any time I wanted to test a new data structure or what have you I knew where everything was. It also (to me at least) makes the traversal algorithms a lot clearer.

I will simplify the implementations as you've suggested. If you do want to get rid of them, let me know.

davidchambers replied:
Your reasoning is sound. Let's keep the aliases. :)

davidchambers commented on index.js:
```diff
@@ -125,6 +136,463 @@
+        }
+      }
+      return null;
+    })(startNode, []);
```
I find `return (function() {})();` a bit awkward. Let's define the function then `return recur(startNode, []);`.

davidchambers commented on index.js:
```diff
@@ -125,6 +136,463 @@
+    acc = f(acc, node, parent);
+    var nodes = arrayClone(node.children);
+    for (var idx = 0; idx < nodes.length; idx += 1) {
+      acc = traverseWith(f, acc, nodes[idx], node);
```
I like to name recursive functions. My reasoning is that a function should depend on the outside world in as few ways as possible. The name of the identifier to which a function is bound is part of the world outside the function. The function's name, though, is part of the function's world.

I'm happy with either of the following names:

```javascript
var traverseWith = function traverseWith(f, acc, node, parent) { /* ... */ };
```

```javascript
var traverseWith = function recur(f, acc, node, parent) { /* ... */ };
```

davidchambers commented on index.js:
```diff
@@ -125,6 +136,463 @@
+    }
+  };
+
+  //  addSeparator :: String -> (Node -> [Node]) -> Void
```
Do you mean `(Node, [Node])` rather than `(Node -> [Node])`?

davidchambers commented on index.js:
```diff
@@ -125,6 +136,463 @@
+    ]);
+  };
+
+  //  Constraints :: {k: Constraint} -> Node
```
Let's use `StrMap Constraint` here.

davidchambers commented on index.js:
```diff
@@ -125,6 +136,463 @@
+  //  Constraints :: {k: Constraint} -> Node
+  var Constraints = function(data) {
+    var tree = Node('CONSTRAINTS', '',
+                     chain(toPairs(data), TypeVarConstraints));
```
:arrow_left:

davidchambers commented on index.js:
```diff
@@ -720,8 +1224,15 @@
+      return chain(commonTypes(map(values, determineActualTypes)),
+                   rejectInconsistent);
+    };
+
```
Thank you for defining this function! We use this expression in quite a few places.

davidchambers commented on index.js:
```diff
@@ -720,8 +1224,15 @@
       });
     };
 
+    //  valueTypes :: [a] -> [Type]
```
The argument's type should be `[Any]` rather than `[a]`, since the values needn't all be members of a common type.

davidchambers commented on index.js:
```diff
@@ -750,16 +1263,37 @@
+                                               merge({
+                                                 indexes: [index],
+                                                 values: [values]
+                                               }));
```
This indentation looks a bit strange to me. I'd prefer:

```javascript
$typeVarMap[typeVarName] =
map(okTypes, merge({indexes: [index], values: [values]}));
```

Indent the second line if you like. :)

kedashoe replied:
double check this one for me after I push to make sure I got it right

davidchambers commented on test/error-messages/nested-invalid-value.txt:
```diff
@@ -0,0 +1,9 @@
+
+1)  "ab" :: String
+
+The value at position 1 is not a member of (Array Number).
```
Were we going to use `???` here in place of `Number`?

kedashoe replied:
I was planning to look into this today. Do you see a very simple way to know when we should do this? My first impression was that this would not be trivial..

davidchambers replied:
This change is not crucial. Unless there's a straightforward solution, let's not worry about it.

kedashoe replied:
Ok I think this did turn out to be simpler than I anticipated (you already had a function dealing with this output, and it was only being called once).

One thing, the `???` seemed a bit conspicuous to me. Kinda the whole point of this was that it doesn't matter and the `???` draws my attention. Instead, we just always return the type variable, so the above error message becomes
```
Invalid value

concat :: Array a -> Array a -> Array a
                     ^^^^^^^
                        1

1)  "ab" :: String

The value at position 1 is not a member of (Array a).
```
and another example from the original test suite
```
Invalid value

unnest :: Array (Array a) -> Array a
          ^^^^^^^^^^^^^^^
                 1

1)  [1, 2, 3] :: (Array Number)

The value at position 1 is not a member of (Array (Array a)).
```

davidchambers replied:
> [T]his did turn out to be simpler than I anticipated[.]

Great!

> [T]he `???` seemed a bit conspicuous to me. Kinda the whole point of this was that it doesn't matter and the `???` draws my attention. Instead, we just always return the type variable[.]

Sounds good to me. :)

davidchambers commented on test/index.js:
```diff
@@ -1253,3 +1264,97 @@ describe('def', function() {
+    throws(function() { concat([1, 2], ['a', 'b']); },
+           errorEq(TypeError, expectedError('nested-type-var-conflict')));
+  });
+});
```
Insert an empty line between the two lines above, please. :)

davidchambers commented on test/index.js:
```diff
@@ -1253,3 +1264,97 @@ describe('def', function() {
+  var expectedError = function(basename) {
+    var path = __dirname + '/error-messages/' + basename + '.txt';
+    return fs.readFileSync(path).toString();
+  };
```
Do you intend to keep this function? In other words, do you think having the error messages in separate text files is cleaner than concatenating string literals in situ?

kedashoe replied:
Not sure. I like them, except having to come up with a name for the files :pensive:  Do you have a vim plugin that auto formats those multi line string concatenations? Always frustrating for me when something on the first line changes and it becomes a cascading game of wack a mole. I guess since these are broken up over multiple lines that actually wouldn't happen. What do you think?

davidchambers replied:
It's very nice to see examples of the messages without the noise of many `'` and `+` characters, but I don't like having "input" and "output" in separate files. Let's go with the (ugly) concatenated strings, if you don't mind.

kedashoe commented:
:zap: + :speech_balloon:

davidchambers commented on index.js:
```diff
@@ -125,6 +136,459 @@
+    return Array(times + 1).join(s);
+  };
+
+  //  findNodePosition :: Node -> ((Node, [Nodes]) -> Node) -> Position
```
If `sig :: a`, we must have `fn :: a -> b` since we have `fn(sig)`. The first `a` is `Node` and the second is `(Node, [Nodes])`, so something is amiss.

davidchambers commented on index.js:
```diff
@@ -125,6 +136,459 @@
+    return sig.children[SIG_IDX_NAME].children[0].text;
+  };
+
+  //  findConstraint :: (TypeClass, String) -> Node -> Node
```
Is the return type `Nullable Node` rather than `Node`?

davidchambers commented on index.js:
```diff
@@ -125,6 +136,459 @@
+    };
+  };
+
+  //  findTypeVar :: (String, Node) -> Node
```
Is the return type `Nullable Node` rather than `Node`?

davidchambers commented on index.js:
```diff
@@ -125,6 +136,459 @@
+    return args[index];
+  };
+
+  //  findArgTypeVar :: (Integer, String) -> Node -> Node
```
Is the return type `Nullable Node` rather than `Node`?

davidchambers commented on index.js:
```diff
@@ -125,6 +136,460 @@
+    return stripNamespace(constraint.name);
+  };
+
+  //  traverse :: ((Node, [Node]) -> Void) -> Node -> Void
```
Let's use [`Undefined`][1] rather than `Void`.


[1]: https://github.com/plaid/sanctuary-def#undefined

davidchambers commented on index.js:
```diff
@@ -125,6 +136,460 @@
+        f(node, parents);
+        var nodes = arrayClone(node.children);
+        for (var idx = 0; idx < nodes.length; idx += 1) {
+          recur(nodes[idx], parents.concat(node));
```
I'd prefer `parents.concat([node])`.

davidchambers commented on index.js:
```diff
@@ -125,6 +136,460 @@
+      }
+      var nodes = arrayClone(node.children);
+      for (var idx = 0; idx < nodes.length; idx += 1) {
+        r = recur(nodes[idx], parents.concat(node));
```
As above, I'd prefer `parents.concat([node])`.

davidchambers commented on index.js:
```diff
@@ -125,6 +136,460 @@
+  };
+
+  //  reject :: ([a], (a -> Boolean)) -> [a]
+  var reject = function(a, pred) {
```
Let's name the first parameter `xs` rather than `a`.

kedashoe replied:
- void -> undefined
- a -> xs
- concating []

:zap:

davidchambers commented on test/index.js:
```diff
@@ -338,39 +359,95 @@ describe('def', function() {
                    '‘$9’ expected a value of type Number ' +
                    'as its seventh argument; received "X"'));
-
+    
```
:space_invader:

davidchambers commented on test/index.js:
```diff
@@ -551,34 +665,75 @@ describe('def', function() {
+                   'length :: { end :: { x :: Number, y :: Number }, start ' +
+                   ':: { x :: Number, y :: Number } } -> Number\n' +
+                   '          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^' +
+                   '^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n' +
```
At first glance I thought we were one caret short. Let's make this tiny change:

```diff
                    'length :: { end :: { x :: Number, y :: Number }, start ' +
                    ':: { x :: Number, y :: Number } } -> Number\n' +
-                   '          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^' +
-                   '^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n' +
+                   '          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^' +
+                   '^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n' +
```

By moving one caret to the second line the last caret lines up with the closing squiggly bracket.

davidchambers commented on test/index.js:
```diff
@@ -889,29 +1234,61 @@ describe('def', function() {
   });
 
-  it('supports the "PositiveInteger" type', function() {
+  it.only('supports the "PositiveInteger" type', function() {
```
Has the time come to remove this? :)

kedashoe replied:
Not quite done with all the error messages yet, was using `.only` to process a block at a time.

davidchambers replied:
No worries. I wasn't sure of the status.

davidchambers commented:
Thank you very much for all your work on this pull request, Kevin!

There are a few remaining tasks to complete before we merge this:

  - address the minor issues I just raised;
  - trim any trailing whitespace;
  - remove the text files (which were very helpful!);
  - rebase (and resolve any merge conflicts); and
  - get `make test lint` to pass.

kedashoe commented:
:zap: , except for rebase, as coverage does not *quite* pass:

- The `null` branch of `head` and `last` are never hit. I could change the signature to the other alternative you suggested, `head :: [a] -> a?`
- The `nest` function passed to `_satisfactoryTypes` is not used anymore, so [this line](https://github.com/plaid/sanctuary-def/pull/23/files#diff-168726dbe96b3ce427e7fedce31bb0bcR1308) and the sibling one just below are never called. I think it is safe to get rid of this parameter, ya?

I've made those changes on my local and coverage is at 100% (and all tests pass)

kedashoe commented:
I've rebased and updated the PR with the coverage changes mentioned above.

See also [this variable swapping](https://github.com/plaid/sanctuary-def/pull/23/files#diff-168726dbe96b3ce427e7fedce31bb0bcR1129) to fix the error I mentioned last night when arguments are applied out of order.

I think the rebase conflicts were pretty simple, just had to wrap some of my changes in the test check from #28, and got rid of a lot of the error messages I updated because of #27

davidchambers commented:
Very exciting! I'll have another look now. :)

davidchambers commented:
Let's apply this patch to avoid the variable swapping:

```diff
diff --git a/index.js b/index.js
index e42f721..78a6356 100644
--- a/index.js
+++ b/index.js
@@ -1122,23 +1122,17 @@
     return num + ')  ' + show(value) + ' :: ' + types.join(', ') + '\n\n';
   };
 
-  var conflictingTypeVar = function(sig, typeVarName,
-                                    prevIndex, prevValue, prevTypes,
-                                    badIndex, badValue, badTypes) {
-    if (prevIndex > badIndex) {
-      var tmpIndex = prevIndex; prevIndex = badIndex; badIndex = tmpIndex;
-      var tmpValue = prevValue; prevValue = badValue; badValue = tmpValue;
-      var tmpTypes = prevTypes; prevTypes = badTypes; badTypes = tmpTypes;
-    }
+  var conflictingTypeVar = function(sig, typeVarName, a, b) {
+    var pair = a.index > b.index ? [b, a] : [a, b];
     return new TypeError(
       'Type-variable constraint violation\n\n' +
       showSignature(sig) + '\n' +
       sigHighlights(sig, [
-        [true, findArgTypeVar(prevIndex, typeVarName)],
-        [true, badTypeVar(badIndex, typeVarName)]
+        [true, findArgTypeVar(pair[0].index, typeVarName)],
+        [true, badTypeVar(pair[1].index, typeVarName)]
       ]) + '\n\n' +
-      showVal(1, prevValue, prevTypes) +
-      showVal(2, badValue, badTypes) +
+      showVal(1, pair[0].value, pair[0].types) +
+      showVal(2, pair[1].value, pair[1].types) +
       'Since there is no type of which both values are members, the ' +
       'type-variable constraint has been violated.\n'
     );
@@ -1270,12 +1264,12 @@
                 throw conflictingTypeVar(
                   signature(name, constraints, expArgTypes, expRetType),
                   typeVarName,
-                  head(prevType.indexes),
-                  head(prevValues),
-                  valueTypes(prevValues),
-                  index,
-                  values[0],
-                  valueTypes(values)
+                  {index: head(prevType.indexes),
+                   value: head(prevValues),
+                   types: valueTypes(prevValues)},
+                  {index: index,
+                   value: values[0],
+                   types: valueTypes(values)}
                 );
               }
             } else {
```

Feel free to use more descriptive names than `a`, `b`, and `pair`. :)

davidchambers commented on index.js:
```diff
@@ -125,6 +136,460 @@
+  //  head :: [a] -> a?
+  var head = function(xs) {
+    return xs[0];
+  };
```
It seems nowhere are we providing `head` as an argument to a higher-order function. In some places we write `xs[0]`; in other places `head(xs)`. I suggest using the former everywhere, even though we break `head`|`last` symmetry.

davidchambers commented on index.js:
```diff
@@ -125,6 +136,460 @@
+  //  tail :: [a] -> [a]
+  var tail = function(xs) {
+    return xs.slice(1);
+  };
```
This function is referenced just once, in `rootChildrenSeparator`. Let's replace `tail(children)` with `children.slice(1)`.

davidchambers commented on index.js:
```diff
@@ -125,6 +136,460 @@
+  };
+
+  //  getParent :: [Node] -> Nullable Node
+  var getParent = last;
```
This surprises me. I expect the type to be `Node -> Nullable Node`. Perhaps `chooseParent` is a better name?

davidchambers commented on index.js:
```diff
@@ -125,6 +136,460 @@
+  var getParent = last;
+
+  //  isRoot :: [Node] -> Boolean
+  var isRoot = isEmpty;
```
As with `getParent`, I'd expect to apply this function to a value of type `Node` rather than `[Node]`.

kedashoe replied:
There is actually even more here to not expect :thought_balloon:

`isRoot` should really be called `isThisWhereTraversalStarted`. I arrived here more or less this way:

1. Keep node creation and representation as simple as possible
2. No pointer to parent
3. We need to know who our parent is and sometimes more distant ancestors when we traverse
4. Maintain running list of ancestors as we traverse

All our tree building is done by each section of the signature, ie we start the traversals from `Name` or `Constraints`, not `signature`. I like this, as it keeps the logic closer to who should know about it and who it affects (note, having parent pointers would not change this, so this need not influence that decision). When these traversal callback functions ask `isRoot`, they are actually asking if we are at `Name` or `Constraints`, not if we are at `signature`.

Long story not quite as long, I don't know where you consider this on the hack scale, but I am happy to refactor the tree to include parent pointers which would result is probably more traditional helper functions.

kedashoe replied:
One other question. For type variable constraint violations in which there are 3+ instances of the type variable, would you rather reference

- the offending occurrence and the first occurrence
- the offending occurrence and the previous occurence

eg

```
Type-variable constraint violation

concat :: Array a -> Array a -> Array a
                ^                     ^
                1                     2

1)  1 :: Number

2)  "bad" :: String

Since there is no type of which both values are members, the type-variable constraint has been violated.
```
vs
```
Type-variable constraint violation

concat :: Array a -> Array a -> Array a
                           ^          ^
                           1          2

1)  1 :: Number

2)  "bad" :: String

Since there is no type of which both values are members, the type-variable constraint has been violated.
```
The code as it currently stands was actually designed to highlight every occurrence, we are just grabbing the head to highlight the first. We can simplify things a bit by picking one or the other (and simplest would be option 2, the previous.. but this is pretty simple either way).

davidchambers replied:
Thanks for the explanation. Now that I know that `[Node]` represents a path, the type signature makes sense to me. :)

As for the question of which type variables to underline, the answer may be the offending occurrence and *all* previous occurrences, since it's the set of types common to all previous `a` values which constrain subsequent `a` values.

>     concat :: Array a -> Array a -> Array a
>                     ^                     ^
>                     1                     2
>
>     1)  1 :: Number
>
>     2)  "bad" :: String

How can we represent the `a` at position 1 (or any other position) as a single value? Shouldn't we list all the array's values? This may be too wordy, so perhaps we should display the common types but omit the values?

kedashoe replied:
> the answer may be the offending occurrence and all previous occurrences, since it's the set of types common to all previous a values which constrain subsequent a values.

Okay, hopefully that won't be too hard since that is what I originally set out to do.

> How can we represent the a at position 1 (or any other position) as a single value? Shouldn't we list all the array's values?

Good questions, the current result is somewhat awkward. I feel like the value could help a developer quickly identify exactly what has gone wrong, seems a shame to lose it. It is just a string, maybe we can truncate it? Still have to figure out how we'd highlight it if we're going to show multiple values. Maybe a special case for "collection" (Array, StrMap) types? Gotta run, will check back later.

davidchambers replied:
> Maybe a special case for "collection" (Array, StrMap) types?

We could list each value on a separate line. If the first argument is a list of integers from 1 to 100, say, and the second is `['foo', 'bar', 'baz']`, we could state:

    concat :: Array a -> Array a -> Array a
                    ^          ^
                    1          2

    1)  1 :: Number, Integer
        2 :: Number, Integer
        ...
        99 :: Number, Integer
        100 :: Number, Integer

    2)  "foo" :: String
        "bar" :: String
        "baz" :: String

    Since there is no type of which all the above values are members, the type-variable constraint has been violated.

I'm not sure whether the `...` should be literal or whether we should display all the value–types pairs regardless of how many there are.

kedashoe replied:
I really like what you've done there with the literal `...`. I look at it messages should convey as much info as simply as possible. I would think in the vast majority of cases, there wouldn't be much to gain from seeing every element of a collection for these types of errors. It would be a bit annoying to have your screen flooded for large collections forcing you to scroll around to see what is going on. There could be a `verbose` flag to show all of every value.

I'll take a crack at this.

kedashoe replied:
Still a wip, but pushed up some changes. Mostly so you can review the error messages. I've added a new `it` block in the test file.

Is it okay to move the error functions inside the `create` closure? It would help to be able to access `determineActualTypes`.

Gotta run, sorry for abbreviated message, will be back later

davidchambers replied:
> Is it okay to move the error functions inside the `create` closure? It would help to be able to access `determineActualTypes`.

Could we instead move `determineActualTypes` out of `create`? It needs access to the environment, so that would need to be provided as an argument. This is not too much of an inconvenience if one makes it a curried function:

```javascript
//  determineActualTypes :: [Type] -> a -> [Type]
var determineActualTypes = function(env) {
  return function recur(value) {
    // ...
  };
};
```

One must then update call sites. For example:

```diff
-                      commonTypes(map(values, determineActualTypes))
+                      commonTypes(map(values, determineActualTypes(env)))
```

kedashoe replied:
> Could we instead move determineActualTypes out of create? It needs access to the environment, so that would need to be provided as an argument. This is not too much of an inconvenience if one makes it a curried function:

I considered this a while ago for another issue, but it won't help for my current problem. Previously, types had been determined and passed as arguments to the error functions. This always bothered me a little but needed to be done because of where `determineActualTypes` was. Now it is not so simple as passing the types since we display things differently for collections. To solve this I pass the `valueTypes` function (which itself calls `determineActualTypes`) as an argument. So I would instead either need to pass around a curried function or `env` itself. Seems a bit awkward to always be passing this same function to all of our error functions. But it works, so not a big deal if you like the error functions where they are.

davidchambers replied:
I'll defer to you on this, Kevin. Go with whichever approach you consider preferable.

kedashoe replied:
My only concern is performance, I don't know how much adding those functions to the closure might slow things down. Since it is already working as is, I figure I'll leave it. It can always be refactored later :)

kedashoe replied:
I feel like we need to wrap these values (especially with StrMap on the way). What do you think of something like this
```js
fromMaybe(0, [1, 2, 3, 4, 5]);
```
```
fromMaybe :: a -> Maybe a -> a
                  ^^^^^^^
                     1

1)  [ 1 :: Number
      2 :: Number
      ...
      4 :: Number
      5 :: Number ]

The value at position 1 is not a member of (Maybe a).
```
and presumably in the future
```js
fromMaybe(0, {a: 1, b: 2, c: 3, d: 4, e: 5});
```
```
fromMaybe :: a -> Maybe a -> a
                  ^^^^^^^
                     1

1)  { a: 1 :: Number
      b: 2 :: Number
      ...
      d: 4 :: Number
      e: 5 :: Number }

The value at position 1 is not a member of (Maybe a).
```

davidchambers replied:
This seems like too much of a special case to me. I'd find this clearer:

```
fromMaybe :: a -> Maybe a -> a
                  ^^^^^^^
                     1

1)  [1, 2, 3, 4, 5] :: (Array Number)

The value at position 1 is not a member of (Maybe a).
```

```
fromMaybe :: a -> Maybe a -> a
                  ^^^^^^^
                     1

1)  {"a": 1, "b": 2, "c": 3, "d": 4, "e": 5} :: Object, (StrMap Number)

The value at position 1 is not a member of (Maybe a).
```

Are you worried that the string representation of a large array may be hundreds or thousands of characters in length?

davidchambers commented on index.js:
```diff
@@ -646,28 +1125,85 @@
+      ]) + '\n\n' +
+      showErrorValue(1, typeFn, value) +
+      'The value at position 1 is not a member of ' +
+      expTypes.join(', ') + '.\n'
```
I think `' or '` would make a better separator.

    The value at position 1 is not a member of Number or Integer

We could be fancy and use the `A`, `A or B`, `A, B, or C` format:

```javascript
//  oxford :: [Any] -> String
var oxford = function(xs) {
  return xs.length <= 2 ?
    xs.join(' or ') :
    xs.slice(0, -1).join(', ') + ', or ' + xs[xs.length - 1];
};
```

davidchambers commented on test/index.js:
```diff
@@ -410,13 +520,203 @@ describe('def', function() {
+                   '\n' +
+                   'Since there is no type of which all the above values ' +
+                   'are members, the type-variable constraint has been ' +
+                   'violated.\n'));
```
This looks great to me!

davidchambers commented on test/index.js:
```diff
@@ -410,13 +520,203 @@ describe('def', function() {
+                   'The value at position 1 is not a member of (Array a).\n'));
+
+    throws(function() { a1([1, 2, 3, 4, 5], ['a', 'b', 'c', 'd', 'e']); },
+      errorEq(TypeError,
```
:triangular_ruler:

davidchambers commented on test/index.js:
```diff
@@ -410,13 +520,203 @@ describe('def', function() {
+                   'violated.\n'));
+
+    throws(function() { a1([[1, 2], [3, 4]], [['a', 'b'], ['c', 'd']]); },
+      errorEq(TypeError,
```
:triangular_ruler:

davidchambers commented on test/index.js:
```diff
@@ -410,13 +520,203 @@ describe('def', function() {
+                   '(Array a)).\n'));
+
+    throws(function() { a2([1, 2, 3, 4, 5], ['a', 'b', 'c', 'd', 'e']); },
+      errorEq(TypeError,
```
:triangular_ruler:

davidchambers commented on test/index.js:
```diff
@@ -410,13 +520,203 @@ describe('def', function() {
+                   '(Array a)).\n'));
+
+    throws(function() { a2([[1, 2], [3, 4]], [['a', 'b'], ['c', 'd']]); },
+      errorEq(TypeError,
```
:triangular_ruler:

davidchambers commented on test/index.js:
```diff
@@ -506,9 +820,18 @@ describe('def', function() {
+                   '| "hours") -> ValidDate -> ValidNumber\n' +
+                   '             ' +
+                   '^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n' +
+                   '                                     1\n' +
```
I suggest we ignore our maximum line length for these errors messages.

```javascript
                   'convertTo :: ("milliseconds" | "seconds" | "minutes" | "hours") -> ValidDate -> ValidNumber\n' +
                   '             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n' +
                   '                                     1\n' +
```

I'd like to be able to confirm visually that the correct characters are underlined (even if I must scroll to do so).

davidchambers commented on index.js:
```diff
@@ -720,11 +1271,81 @@
+      );
+    };
+
+    //  typesAndValues :: [Any] -> ErrorRepr
```
`s/typesAndValues/nullaryTypesAndValues/`

davidchambers commented on test/index.js:
```diff
@@ -445,7 +805,27 @@ describe('def', function() {
+    var anonJust = function(x) {
+      return {
+        '@@type': 'my-package/anonMaybe',
+        'constructor': anonJust,  // ramda/ramda#1516
```
Let's name the function `anonJust` (in addition to assigning it to a variable of that name). It doesn't seem right to reference a variable in outer scope (which does not even have the intended value at the time of this function's creation). Self-referential functions should be named functions, I believe.

davidchambers commented on test/index.js:
```diff
@@ -445,7 +805,27 @@ describe('def', function() {
   it('supports custom types', function() {
-    var env = $.env.concat([Integer, $Pair]);
+
+    //  anonJust :: a -> Just a
```
Should the type be `a -> Maybe a` rather than `a -> Just a`? Also, should we name the function `AnonJust` rather than `anonJust`, to follow the convention of capitalizing the names of data constructors? Similarly, should we name the type `AnonMaybe`?

davidchambers commented on test/index.js:
```diff
@@ -479,8 +860,37 @@ describe('def', function() {
+    var fromMaybe = def('fromMaybe', {}, [a, anonMaybe(a), a],
+                                         function(x, maybe) {
+      return maybe.isJust ? maybe.value : x; }
+    );
```
Let's go with this formatting:

```javascript
//  fromMaybe :: a -> anonMaybe a
var fromMaybe =
def('fromMaybe',
    {},
    [a, anonMaybe(a), a],
    function(x, maybe) { return maybe.isJust ? maybe.value : x; });
```

When line wrapping of arguments is required, I like each argument to begin at the same column position. It makes it easier to see how things are nested.

davidchambers commented on index.js:
```diff
@@ -672,32 +1129,94 @@
+      LEFT_SINGLE_QUOTATION_MARK + sigGetName(sig) +
+      RIGHT_SINGLE_QUOTATION_MARK + ' requires ' + LEFT_SINGLE_QUOTATION_MARK +
+      typeVarName + RIGHT_SINGLE_QUOTATION_MARK + ' to satisfy the ' +
+      typeClass + ' type-class constraint; The value at position 1 does not.\n'
```
`s/The/the/`

davidchambers commented on index.js:
```diff
@@ -835,7 +1441,7 @@
         var $typeVarMap = {};
         for (var typeVarName in _typeVarMap) {
-          $typeVarMap[typeVarName] = _typeVarMap[typeVarName].slice();
+          $typeVarMap[typeVarName] = _typeVarMap[typeVarName];
```
@kedashoe and I discussed this on Gitter and agreed that cloning the arrays has never been necessary. My guess is that while working on the initial version I used `R.clone`, and later, when I removed the Ramda dependency, wrote this code to perform a "deep" clone manually.

davidchambers commented on test/index.js:
```diff
@@ -445,7 +805,27 @@ describe('def', function() {
+    var AnonJust = function(x) {
+      return {
+        '@@type': 'my-package/AnonMaybe',
+        'constructor': AnonJust,  // ramda/ramda#1516
```
`s/AnonJust/function() {}/`

kedashoe commented:
:zap: :zap:

davidchambers commented on index.js:
```diff
@@ -69,6 +69,18 @@
     };
   };
 
+  //  functionName :: f -> String
```
This type is actually `Function -> String`.

davidchambers commented on index.js:
```diff
@@ -69,6 +69,18 @@
+    return match == null ? '' : match[1];
+  };
+
+  //  constructorName :: x -> String
```
I think the type is best represented as `{ constructor :: Function } -> String`.

davidchambers commented on index.js:
```diff
@@ -641,6 +1090,14 @@
+      LEFT_SINGLE_QUOTATION_MARK + x + RIGHT_SINGLE_QUOTATION_MARK
+    );
+  };
+
```
Let's move this definition above the first reference to `unexpectedType`.

kedashoe commented:
:zap: :zap: :zap:

davidchambers commented on test/index.js:
```diff
@@ -74,7 +76,7 @@ var Nothing = function Nothing() {
 };
 
-//  Just :: a -> Just a
+//  Just :: a -> Maybe a
```
Thanks for correcting this error!

davidchambers commented:
:white_large_square: :deciduous_tree: :deciduous_tree: :deciduous_tree: :white_large_square:
:deciduous_tree: :deciduous_tree: :deciduous_tree: :deciduous_tree: :deciduous_tree:
:deciduous_tree: :deciduous_tree: :deciduous_tree: :deciduous_tree: :deciduous_tree:
:deciduous_tree: :deciduous_tree: :deciduous_tree: :deciduous_tree: :deciduous_tree:
:deciduous_tree: :deciduous_tree: :deciduous_tree: :deciduous_tree: :deciduous_tree:
:white_large_square: :deciduous_tree: :deciduous_tree: :deciduous_tree: :white_large_square:
:white_large_square: :white_large_square: :door: :white_large_square: :white_large_square:

benperez commented:
> :white_large_square: :deciduous_tree: :deciduous_tree: :deciduous_tree: :white_large_square:
:deciduous_tree: :deciduous_tree: :deciduous_tree: :deciduous_tree: :deciduous_tree:
:deciduous_tree: :deciduous_tree: :deciduous_tree: :deciduous_tree: :deciduous_tree:
:deciduous_tree: :deciduous_tree: :deciduous_tree: :deciduous_tree: :deciduous_tree:
:deciduous_tree: :deciduous_tree: :deciduous_tree: :deciduous_tree: :deciduous_tree:
:white_large_square: :deciduous_tree: :deciduous_tree: :deciduous_tree: :white_large_square:
:white_large_square: :white_large_square: :door: :white_large_square: :white_large_square:

The elusive mega-tree.

kedashoe commented:
:tada: :beers:

davidchambers commented:
> The elusive mega-tree.

It's almost as rare as the dromedary of dromedaries.
